libexec/rtld-elf/rtld_c18n.c

```diff
@@ -787,28 +787,29 @@ tramp_hook_impl(void *rcsp, int event, void *target, const Obj_Entry *obj,
 	struct utrace_rtld ut;
 	static const char rtld_utrace_sig[RTLD_UTRACE_SIG_SZ] = RTLD_UTRACE_SIG;
 
-	sym_num = def == NULL ? 0 : def->st_name;
-	sym = def == NULL ? "<unknown>" : strtab_value(obj, def->st_name);
-	callee = comparts.data[obj->compart_id]->name;
+	if (ld_compartment_utrace != NULL) {
+		sym_num = def == NULL ? 0 : def->st_name;
+		sym = def == NULL ? "<unknown>" : strtab_value(obj, sym_num);
+		callee = comparts.data[obj->compart_id].name;
 
 #ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
-	caller_id =
-	    ((uintptr_t *)cheri_setoffset(rcsp, cheri_getlen(rcsp)))[-2];
-	(void)link;
+		(void)link;
 #else
-	if (cheri_gettag(link) &&
-	    (cheri_getperm(link) & CHERI_PERM_EXECUTIVE) == 0)
-		caller_id = ((uintptr_t *)
-		    cheri_setoffset(rcsp, cheri_getlen(rcsp)))[-2];
-	else
-		caller_id = C18N_RTLD_COMPART_ID;
+		if (cheri_gettag(link) &&
+		    (cheri_getperm(link) & CHERI_PERM_EXECUTIVE) == 0) {
+#endif
+			rcsp = cheri_setoffset(rcsp, cheri_getlen(rcsp));
+		        caller_id = rcsp[-2];
+#ifndef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
+		} else
+			caller_id = C18N_RTLD_COMPART_ID;
 #endif
```

Comment:
```C
#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
		(void)link;
#else
		if (cheri_gettag(link) &&
		    (cheri_getperm(link) & CHERI_PERM_EXECUTIVE) == 0)
#endif
		{
			rcsp = cheri_setoffset(rcsp, cheri_getlen(rcsp));
		        caller_id = rcsp[-2];
		}
#ifndef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
		else
			caller_id = C18N_RTLD_COMPART_ID;
#endif
```
Something like this makes my head hurt less

Reply:
Though I don't like the refactoring and bug fixing being in the same commit. Presumably the bug is putting more under the `if (ld_compartment_utrace != NULL)`? But that's got nothing to do with the deduplication of code and changing of types to avoid casts.